question2/odbcdemo.c

```diff
@@ -0,0 +1,288 @@
+#include <sqlext.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <strings.h>
+
+/*
+** Define Some useful defines
+*/
+#if !defined(NULL)
+#define NULL 0
+#endif
+
+/*
+** function: ODBC_error
+**
+** Purpose: Display to stdout current ODBC Errors
+**
+** Arguments: henv _ ODBC Environment handle.
+** hdbc - ODBC Connection Handle error generated on.
+** hstmt - ODBC SQL Handle error generated on.
+**
+** Returns: void
+**
+*/
+void ODBC_error(                 /* Get and print ODBC error messages */
+                SQLHANDLE henv,  /* ODBC Environment */
+                SQLHANDLE hdbc,  /* ODBC Connection Handle */
+                SQLHANDLE hstmt) /* ODBC SQL Handle */
+{
+    UCHAR sqlstate[10];
+    UCHAR errmsg[SQL_MAX_MESSAGE_LENGTH];
+    SQLINTEGER nativeerr;
+    SQLSMALLINT actualmsglen;
+    RETCODE rc;
+loop:
+    rc = SQLError(henv, hdbc, hstmt, (SQLCHAR *)sqlstate, &nativeerr,
+                  (SQLCHAR *)errmsg, SQL_MAX_MESSAGE_LENGTH - 1, &actualmsglen);
+    if (rc == SQL_ERROR) {
+        printf("SQLError failed!\n");
+        return;
+    }
+    if (rc != SQL_NO_DATA_FOUND) {
```

Comment:
there should be a check for SQL_INVALID_HANDLE here, otherwise we would have an infinite loop
```suggestion
    if (rc == SQL_INVALID_HANDLE) {
        printf("Received SQLRETURN= SQL_INVALID_HANDLE!\n");
        return;
    }
    if (rc != SQL_NO_DATA_FOUND) {
```